Add Kanban board widget

trobonox opened the pull request:
![preview](https://user-images.githubusercontent.com/57040351/230682338-fdb1fde4-3994-4ffe-bc65-4ec95fa440ef.png)

Implements the Kanban board widget as discussed and seen in #266 

**Currently known issues:**
- Drag and drop is a bit weird/offset, fix unknown (seems to be because of the draggable container)
- Widget can be dragged outside of the grid's bounds which causes overflow (this issue is caused by the fixed size of draggable widget containers, will be fixed with the PR by @royanger)

**Current Todos**
- [x] Remove extra CSS and use Tailwind --> mostly done, not sure if custom rbd selector can be queried with tailwind
- [x] Remove column border to have more space for cards/prevent cards being squished
- [x] Make widget resizable
- [x] Fix task add bug introduced after rewrite
- [x] Auto scroll to show cancel and add buttons when adding a new task
- [x] Implement card/task editing
- [x] Add myself to contributors.yml
- [ ] Check if removal of IBackgroundColor in store.tsx causes any issues (seems to be a dead import or renamed? hopefully not something that broke while rebasing) 
- [ ] Maybe improve resizing and handling of the minimum height?

Potential future Todos:
- Custom scrollbars on column overflow
- Card descriptions
- Card colors
- Draggable columns
- Ability to add, remove and rename columns
All of these could be added down the line but would need some extra work like a Modal to pop up when clicking on cards to present all of these options.

**Questions for review:**
- Is the way I mutate state fine, or should the addTask and deleteTask functions be removed with appropriate equivalents directly inside the state?
- Anything else that would need to be changed in the implementation?

trobonox commented:
This is kind of done, ready to go into the first review round I'd say since there's quite a bit to look at.
Not sure what to do about those merge conflicts, let me know if there's something I can do from my side.

royanger commented:
Are you new to dealing with merge conflicts and doing rebases?

Melkeydev commented on src/components/Kanban/kanban.css:
```diff
@@ -0,0 +1,13 @@
+@media (min-width: 641px) {
```
For consistency - can we stick to using tailwind?

trobonox replied:
I'll try to reduce the CSS to a minimum, but:
I haven't found a tailwind breakpoint for the exact width you're using for Desktop and I hesitated to edit the tailwind config. If you believe this could be useful, I could go ahead and add a custom "desktop" breakpoint in the config.
The other part of the CSS is a needed "hack" to fix drag and drop just on Desktop sizes and I'm not sure if I can query the selector with tailwind. (the [data-rbd-draggable-id] selector)
I'll look into it to see when that's applied.

royanger replied:
The fix for drag and drop (where part will go past the edge) is coming with the Hacker News PR. You can prep for it by adding `width=<whatever>` (Ie `width="48rem"`) as a prop to the DWrapper component wrapping your Kanban in Astrostation.tsx

As for the Tailwind breakpoint, its 'sm'. (Yes, its 640px vs 641px).

trobonox replied:
I see, that's very useful info, thanks.
I should be able to remove most CSS then.

The only remaining part would be [data-rbd-draggable-id] and it's mostly to prevent elements from jumping around when dragging them and I haven't found another fix for this (and I'm not sure if your hacker news PR would fix that too since it seems to be more of an issue caused by nested draggables, see the currently known issues)

royanger replied:
I have no internal drags for the Hacker News PR, no nothing toughing that.

Melkeydev commented:
![image](https://user-images.githubusercontent.com/53410236/231044311-2da465d2-9b40-4509-989c-1185946fb123.png)
@trobonox - do you have any ideas of how to handle this overflow? it looks a bit.. strange.
Also - do you think resizing can be useful for something like this? I can totally seeing it getting kind of squished..

royanger commented:
> ![image](https://user-images.githubusercontent.com/53410236/231044311-2da465d2-9b40-4509-989c-1185946fb123.png) @trobonox - do you have any ideas of how to handle this overflow? it looks a bit.. strange. Also - do you think resizing can be useful for something like this? I can totally seeing it getting kind of squished..

There is a path to resolution suggested in the Issue for this.

trobonox commented:
> Are you new to dealing with merge conflicts and doing rebases?

No, I've done them before. I'm just blind and didn't see the "resolve conflicts" button on this PR so I was confused where to see them.
I'll take a look and try to fix it.

trobonox commented:
> ![image](https://user-images.githubusercontent.com/53410236/231044311-2da465d2-9b40-4509-989c-1185946fb123.png)
> @trobonox - do you have any ideas of how to handle this overflow? it looks a bit.. strange.
> Also - do you think resizing can be useful for something like this? I can totally seeing it getting kind of squished..

So I had two ideas for this: either I could remove the column borders as Roy suggested in the linked issue, or we could work with a thin custom scrollbar so the items don't look that squished when overflow occurs. The first option is probably better but might be optimal paired with the 2nd one (sadly full custom scrollbars are only supported on Chromium-based browsers so it's not a universal solution).

The idea of resizing the widget is really good to prevent things getting squished, is @royanger working on the possibility to do this? Or is there already a way to do so?

trobonox commented:
Updated my todos a little, will work on these in the next few days (edited my initial comment)

trobonox commented:
> The current deployment is showing UUID or CUID ids for the tasks, not the entered text. See screen shot.
> ![image](https://user-images.githubusercontent.com/24333362/231087962-377a5e7d-09b4-4861-bb9a-f5154e955e62.png)

This is part of the new todos I added and probably a bug I introduced yesterday while refactoring. Will take a look at it, the fix should be fairly straightforward.

trobonox commented:
Started working on this again and also rebased :)

Melkeydev reviewed (commented):
For some reason i am unable to branch and check this out locally :thinking:

Melkeydev reviewed (commented):
On Mobile + non desktop, there are some responsiveness issues.
I think due to kanban, we can just disable it if the user is using it on a not acceptable window size

![image](https://github.com/MelkeyOSS/astrostation/assets/53410236/afb2d1cf-0f5c-470d-b1fa-b14a1da2ea14)

Melkeydev commented on src/components/Kanban/Kanban.tsx:
```diff
@@ -0,0 +1,272 @@
+  const [cardInputValue, setCardInputValue] = useState(task.name);
+
+  const enableTaskEdit = () => {
+    setCardInputValue(task.name);
+    setCardEditMode(true);
+  }
```
We should add a rule here that if it is in edit mode, we should not bet able to move it

Melkeydev commented on src/components/Kanban/Kanban.tsx:
```diff
@@ -0,0 +1,272 @@
+
+  const onFormSubmit = (event: FormEvent) => {
+    event.preventDefault();
+
```
nit: remove extra space

Melkeydev commented on src/components/Kanban/Kanban.tsx:
```diff
@@ -0,0 +1,272 @@
+    element?.scrollIntoView({ behavior: "smooth", block: "nearest", inline: "start" });
+  }, []);
+
+
```
nit: remove extra space

Melkeydev commented on src/components/Kanban/Kanban.tsx:
```diff
@@ -0,0 +1,272 @@
+export const Kanban = ({}) => {
+  const { setIsKanbanToggled } = useToggleKanban();
+  const { board, setColumns } = useKanban();
+  const isDesktop = useMediaQuery("(min-width: 641px)");
```
This 641 px is based off of where?

trobonox replied:
Not 100% sure but I think I copied this from Astrostation.tsx on some older commit on the main branch
Should this be changed to another value?

trobonox replied:
Digging through the code a little, it also still seems to be present in `WidgetControlModal.tsx`
Not sure why the values are different, but I changed the one I'm using in the code to the current one from `Astrostation.tsx`.

Melkeydev commented on src/components/Kanban/Kanban.tsx:
```diff
@@ -0,0 +1,272 @@
+  const isDesktop = useMediaQuery("(min-width: 641px)");
+
+  const addTask = (columnIndex: number, taskName: string) => {
+    let columns = board.columns;
+    columns[columnIndex].tasks.push({ id: v4(), name: taskName });
```
Would it possibly make sense to double check if there is an existing task in the kanboard with the same name?

trobonox replied:
If we want names to be unique, we could add that. Totally up to you.
At the moment it could make sense but if we ever added more complex features like descriptions, you could have cards with the same name but different descriptions inside, so not sure if we want to enforce this limit.

trobonox replied:
Thought about it a bit more and imo enforcing this limit doesn't make too much sense and we should just let the user do whatever they want.

Melkeydev commented on src/components/Kanban/Kanban.tsx:
```diff
@@ -0,0 +1,272 @@
+    if (!confirm(`Are you sure you want to delete the task "${columns[columnIndex].tasks[taskIndex].name}"?`))
+      return;
+
+    columns[columnIndex].tasks.splice(taskIndex, 1);
```
I wonder if this can lead to some troublesome behaviour.
What happens at taskIndex == 0?

trobonox replied:
When I was testing things seemed fine, I'll try for this specific case again and see.

trobonox replied:
Tested this and it seems fine. Let me know if I'm missing something in my logic but taskIndex == 0 just means the first task in a column is deleted. If you're concerned about the "1" arg afterwards, that just means that we're explicitly removing only one single element from the array (which in this case would be at index 0, so the first one).

Melkeydev commented on src/components/Kanban/Kanban.tsx:
```diff
@@ -0,0 +1,272 @@
+  const onDragEnd = result => {
+    const { destination, source, draggableId } = result;
+
+    if (!destination) {
+      return;
+    }
+
+    if (destination.droppableId === source.droppableId && destination.index === source.index) {
+      return;
+    }
```
Can probably combine these

Melkeydev reviewed (commented):
Left some comments

trobonox commented:
> On Mobile + non desktop, there are some responsiveness issues. I think due to kanban, we can just disable it if the user is using it on a not acceptable window size
> 
> ![image](https://user-images.githubusercontent.com/53410236/255449067-afb2d1cf-0f5c-470d-b1fa-b14a1da2ea14.png)

Seems good to me, has this been done for any other widgets yet or would I have to implement that from scratch? Just want to make sure I don't implement from scratch but there's already a way you're doing it somewhere else.

trobonox commented:
Will be working on the rest of the stuff tomorrow and then it should be ready for another review

trobonox commented:
For the mobile stuff, I've fixed the mobile version to make it responsive again.
Let me know if we rather want to have a long stacked version of the widget or if disabling it entirely on small screens is still a better approach.

trobonox commented:
This is ready to be reviewed again btw

Melkeydev reviewed (approved):
Lets
Get 
This
Money
